### PR TITLE
8296972: [macos13] java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java: getExtendedState() != 6 as expected.

### PR DESCRIPTION
**Pull request by alisenchung**
added displayChanged call to CPlatformWindow when frame first needs to deiconify or unmaximize
All client tests passed after change

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296972](https://bugs.openjdk.org/browse/JDK-8296972): [macos13] java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java: getExtendedState() != 6 as expected. (**Bug** - P4)

### Reviewers
 * [Dmitry Markov](https://openjdk.org/census#dmarkov) (@dmarkov20 - **Reviewer**)
 * [Harshitha Onkar](https://openjdk.org/census#honkar) (@honkar-jdk - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14226/head:pull/14226` \
`$ git checkout pull/14226`

Update a local copy of the PR: \
`$ git checkout pull/14226` \
`$ git pull https://git.openjdk.org/jdk.git pull/14226/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14226`

View PR using the GUI difftool: \
`$ git pr show -t 14226`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14226.diff">https://git.openjdk.org/jdk/pull/14226.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14226#issuecomment-1568912817)

### CODE REVIEW

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java [hunk 1]
```diff
@@ -986,13 +986,15 @@ public void setWindowState(int windowState) {
                     // let's return into the normal states first
                     // the zoom call toggles between the normal and the max states
                     unmaximize();
+                    peer.displayChanged();
```
Comment (dmarkov20) [inline src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java:989]: Why do you call `peer.displayChanged()` only when the previous windows state is `MAXIMIZED_BOTH`? What happens if the previous windows state is `NORMAL`?

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java [hunk 2]
```diff
@@ -986,13 +986,15 @@ public void setWindowState(int windowState) {
...
                 }
                 execute(CWrapper.NSWindow::miniaturize);
                 break;
             case Frame.MAXIMIZED_BOTH:
                 if (prevWindowState == Frame.ICONIFIED) {
                     // let's return into the normal states first
                     execute(CWrapper.NSWindow::deminiaturize);
+                    peer.displayChanged();
```
Comment (dmarkov20) [inline src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java:997]: Same as above. What happens if we move from `NORMAL` to `MAXIMIZED_BOTH`?
 
Shall we call `peer.displayChanged()` when a frame is moved to `NORMAL` state? I know that cases are not covered by the test but it seems the issue takes place there. Can you investigate, please?

**Review by dmarkov20 — CHANGES_REQUESTED** [review]

**Comment by mrserb** [comment]
Why this method is not called automatically by our native code? what events are changed since macOS 13.0?

**Comment by mrserb** [comment]
BTW it is unclear what the call to "displayChanged" will fix. Will it set the correct "graphicsDevice", if yes then what gD was used before the call, if not then what property is updated by the "displayChanged"?

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java (hunk 2 above)
Comment (alisenchung) [reply src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java:997]: I believe the main issue is that there is a native event that changes the window state, but when two state changes are done back to back sometimes that native event isn't delivered quickly enough before the second state change occurs, which causes this test to fail. When a frame is moved to a normal state, there are no back to back state changes, since all we do is change the state back to normal.

Comment (alisenchung) [reply src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java:997]: If I set a delay between the state changes, the problem will also disappear and the test passes.

**Comment by alisenchung** [comment]
> BTW it is unclear what the call to "displayChanged" will fix. Will it set the correct "graphicsDevice", if yes then what gD was used before the call, if not then what property is updated by the "displayChanged"?

I think the problem is that the first state change (uniconify) causes a native event that changes the frame's state, but that event isn't delivered until after the second state change (maximize) occurs, which is causing problems on the final state of the frame. The displayChanged call between these state changes was intended to flush the native event queue and cause the event to be delivered between the state changes.

**Comment by honkar-jdk** [comment]
> Why this method is not called automatically by our native code? what events are changed since macOS 13.0?

@mrserb There might be a new window state in macOS 13 due to a the new [Stage Manager](https://support.apple.com/en-us/HT213315) feature but I believe this shouldn't affect any existing states.

**Comment by alisenchung** [comment]
> If I got it right the switching to `ICONIFIED`/`MAXIMIZED_BOTH` state takes more time on macOS 13 than on previous macOS version. So the test failure is triggered by the lack of synchronisation. When you added displayChanged() call you introduced some kind of delay and the problem stopped happening but it is still present. I do not think any changes are required in JDK but the test definitely needs some modifications. In particular I believe the test must check the state of the frame only when the corresponding `window state changed` event is received.

I did some testing and I think that without the delay, when the frame moves from iconified to maximized states, the second windowStateChanged event never comes (first being iconified -> normal, second being normal -> maximized), so the test will not pass even if we only check state of the frame on a windowStateChanged event. 

Why do you think no changes are required in the JDK? Shouldn't there be some synchronization needed when calling the native deminiaturize and maximize functions in CPlatformWindow?

**Comment by mrserb** [comment]
Possibly some synchronization is needed in the jdk, it needs to be checked why the message is not posted. But it should not be fixed by the "displayChanged" used as a "delay"

**Comment by honkar-jdk** [comment]
@alisenchung @mrserb 

After reading through the discussion, I think the fix on native side rather than on Java might work better here. I was revisiting a Tray Icon issue which had similar sync problem when I was trying to get updated scale value on Java side by using a DisplayChangedListener [Details here - [PR#8441](https://github.com/openjdk/jdk/pull/8441#issuecomment-1121345834)] and there were multiple events occurring back to back. When I had a hard-coded delay set on Java side, the update to tray icon worked fine. I think this is a similar case where introducing a delay solves the issue. But since hard-coded delay is not the ideal solution probably triggering/ calling the required piece of code after everything is updated on native side is the way to go (event/message-driven).

Handling the code on Window's native code with the help of windows messaging event solved the Trayicon problem.

**Comment by honkar-jdk** [comment]
Also noticed this comment in setWindowState() of [CPlatformWindow.java#L1011 ](https://github.com/openjdk/jdk/blob/ea41907396a70933c4f6e6b0f247f7ba589a3a23/src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java#L1011)

_// NOTE: the SWP.windowState field gets updated to the newWindowState
//       value when the native notification comes to us_

May be waiting untill the required native notification using an event-driven wait might work here?

**Comment by honkar-jdk** [comment]
@alisenchung Before digging deeper on Cocoa Event Handling, please check if fixing on native side by waiting for an approriate NSEvent is a viable solution in this case. 

A gist of Cocoa Event Handling mechanism - 

1. NSRunLoop in the event loop [NSApplicationAWT.m#L303](https://github.com/openjdk/jdk/blob/ea41907396a70933c4f6e6b0f247f7ba589a3a23/src/java.desktop/macosx/native/libosxapp/NSApplicationAWT.m#L303) handled by NSApp. 
2. NSApp converts this event into NSEvent object and dispatches using sendEvent to the appropriate NSView object.

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java [hunk 3]
```diff
@@ -986,13 +986,16 @@ public void setWindowState(int windowState) {
                     // let's return into the normal states first
                     // the zoom call toggles between the normal and the max states
                     unmaximize();
+                    peer.getLWToolkit().realSync();
```
Comment (dmarkov20) [inline src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java:989]: Well.. `realSync()` works here but I am not sure if that invocation really fixes the issue or just hides it by introducing a delay.
I believe we need some function which ensures that the frame is in correct (`NORMAL`) state right after `unmaximize()` call and before `execute(CWrapper.NSWindow::miniaturize)` invocation.

Possible implementation of that method below:

```
private void waitForWindowState(int state) {
        Object lock = new Object();
        target.addWindowStateListener(new WindowStateListener() {
            public void windowStateChanged(WindowEvent e) {
                synchronized(lock) {
                    if (e.getNewState() == state) {
                        lock.notifyAll();
                    }
                }
            }
        });
        if (peer.getState() != state) {
            synchronized(lock) {
                try {
                    lock.wait();
                } catch(InterruptedException ie) {}
            }
        }
    }
```

So in code you will call `waitForWindowState(Frame.NORMAL)` instead of `realSync()`

**Review by dmarkov20 — CHANGES_REQUESTED** [review]

**Comment by honkar-jdk** [comment]
@alisenchung 

The current fix works for the cases listed in the test but fails for the following scenarios. To make the test more stable it is better to reassign the static variable `lastFrameState = Frame.NORMAL` in examineStates after creating a new Frame [here ](https://github.com/openjdk/jdk/blob/57b8251241e2044d5039ce162bf4637a9b2e5466/test/jdk/java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java#L78), so that each time a new case is run it sets lastFrameState to default state.

```
// CASE 1:
examineStates(new int[] {Frame.NORMAL, Frame.MAXIMIZED_BOTH, Frame.ICONIFIED});
// CASE 2:
examineStates(new int[] {Frame.NORMAL, Frame.ICONIFIED, Frame.MAXIMIZED_BOTH});
```

**Review by honkar-jdk — CHANGES_REQUESTED** [review]
@alisenchung 

The current fix works for the cases listed in the test but fails for the following scenarios. To make the test more stable it is better to reassign the static variable `lastFrameState = Frame.NORMAL` in examineStates after creating a new Frame [here ](https://github.com/openjdk/jdk/blob/57b8251241e2044d5039ce162bf4637a9b2e5466/test/jdk/java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java#L78), so that each time a new case is run it sets lastFrameState to default state.

```
// CASE 1:
examineStates(new int[] {Frame.NORMAL, Frame.MAXIMIZED_BOTH, Frame.ICONIFIED});
// CASE 2:
examineStates(new int[] {Frame.NORMAL, Frame.ICONIFIED, Frame.MAXIMIZED_BOTH});
```

**Comment by prrace** [comment]
I don't like the idea of a delay in either of the ways discussed here.
And although I do think that its much better to wait for delivery of an event,  I'm not sure I understand the windowState listener proposal.
How do you know in this code that the transition of
iconified-> normal will then followed by normal->maximised ?
Perhaps you'll wait for ever ?
But I don't know what the macOS native model is for all of this.
This sequence is kind of odd and I'm curious what happens on older macOS releases - Sergey touched on this but I don't see direct followup.
There seems to be assumptions it was like this previously but now it is slower ?

I suggest explicit verification of what happens on macOS 12.6.
Are we now getting two native events when there used to be only one ?

**Comment by alisenchung** [comment]
> I don't like the idea of a delay in either of the ways discussed here. And although I do think that its much better to wait for delivery of an event, I'm not sure I understand the windowState listener proposal. How do you know in this code that the transition of iconified-> normal will then followed by normal->maximised ? 

The code here is trying to do a transition from iconified -> maximized, but splits it up into iconified -> normal and normal -> maximized.

**Comment by dmarkov20** [comment]
> I don't like the idea of a delay in either of the ways discussed here. And although I do think that its much better to wait for delivery of an event, I'm not sure I understand the windowState listener proposal. How do you know in this code that the transition of iconified-> normal will then followed by normal->maximised ? Perhaps you'll wait for ever ? But I don't know what the macOS native model is for all of this. This sequence is kind of odd and I'm curious what happens on older macOS releases - Sergey touched on this but I don't see direct followup. There seems to be assumptions it was like this previously but now it is slower ?
> 
> I suggest explicit verification of what happens on macOS 12.6. Are we now getting two native events when there used to be only one ?

We cannot move a frame directly from `ICONIFIED` to `MAXIMIZED_BOTH` and vice versa. We need to go through the `NORMAL` state, (i.e. `ICONIFIED` -> `NORMAL` -> `MAXIMIZED_BOTH`). If my memory does not fail me we use that approach from the very beginning.

We perform two state-transitions subsequently and don't care of real state of the frame between them. It used to work till macOS 13 where the transition from one state to another takes a bit longer. 
In other words we do ICONIFIED` -> `NORMAL` and then immediately `NORMAL` -> `MAXIMIZED_BOTH` but the frame is NOT in `NORMAL` state yet.

**Review by dmarkov20 — CHANGES_REQUESTED** [review]

File: test/jdk/java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java [hunk 4]
```diff
@@ -45,7 +45,7 @@
 
 public class MaximizedToIconified
 {
-    static volatile int lastFrameState = Frame.NORMAL;
+    static volatile int lastFrameState;
```
[the remark this answers is not in the document]
Comment (honkar-jdk) [reply test/jdk/java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java:48]: @dmarkov20 Although the test does not explicitly test the cases mentioned below, it fails for the following two cases because the lastFrameState is changed as we progress through the states and should be reassigned when testing a new set of states. @alisenchung  is assigning this var within examineStates now.

> @alisenchung
> 
> The current fix works for the cases listed in the test but fails for the following scenarios. To make the test more stable it is better to reassign the static variable `lastFrameState = Frame.NORMAL` in examineStates after creating a new Frame [here ](https://github.com/openjdk/jdk/blob/57b8251241e2044d5039ce162bf4637a9b2e5466/test/jdk/java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java#L78), so that each time a new case is run it sets lastFrameState to default state.
> 
> ```
> // CASE 1:
> examineStates(new int[] {Frame.NORMAL, Frame.MAXIMIZED_BOTH, Frame.ICONIFIED});
> // CASE 2:
> examineStates(new int[] {Frame.NORMAL, Frame.ICONIFIED, Frame.MAXIMIZED_BOTH});
> ```

[the remark this answers is not in the document]
Comment (dmarkov20) [reply test/jdk/java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java:48]: Ah, I see now. I would suggest including that cases into the test.

File: test/jdk/java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java [hunk 5]
```diff
@@ -114,6 +116,7 @@ private static void doTest() {
         //    because Toolkit.isFrameStateSupported() method reports these states
         //    as not supported. And such states will simply be skipped.
         examineStates(new int[] {Frame.MAXIMIZED_BOTH, Frame.ICONIFIED, Frame.NORMAL});
+        System.out.println("------");
         examineStates(new int[] {Frame.ICONIFIED, Frame.MAXIMIZED_BOTH, Frame.NORMAL});
```
Comment (honkar-jdk) [inline test/jdk/java/awt/Frame/MaximizedToIconified/MaximizedToIconified.java:121]: @alisenchung Now that the `lastFrameState` is fixed, I would suggest adding the cases mentioned [here](https://github.com/openjdk/jdk/pull/14226/files#r1237259547) too, to make the test comprehensive.

**Review by honkar-jdk — COMMENTED** [review]
The update fix - with the current state check before acquiring the lock and removing the WindowStateListener suggested by @dmarkov20 avoids unnecessary acquiring of lock  and looks to be a better solution that before.

**Review by honkar-jdk — APPROVED** [review]

**Review by dmarkov20 — COMMENTED** [review]
@alisenchung The fix looks OK. Did you test it on macOS 12? If not, can you ensure that previous macOS version(s) works fine with the proposed changes?

**Comment by alisenchung** [comment]
> @alisenchung The fix looks OK. Did you test it on macOS 12? If not, can you ensure that previous macOS version(s) works fine with the proposed changes?

I've tested on macOS12 and it works fine with the changes.

**Review by dmarkov20 — APPROVED** [review]

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java [hunk 6]
```diff
@@ -965,6 +966,33 @@ public boolean isFullScreenMode() {
...
+                    }
+                }
+            }
+        };
+
+        target.addWindowStateListener(wsl);
+        if (peer.getState() != state) {
+            synchronized (lock) {
+                try {
+                    lock.wait();
```
Comment (mrserb) [inline src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java:989]: Are we sure that waiting forever is a good thing? Can this request be ignored by the OS for any reason?

Comment (dmarkov20) [reply src/java.desktop/macosx/classes/sun/lwawt/macosx/CPlatformWindow.java:989]: It is very unlikely that we are waiting forever here unless there are some serious problems with corresponding platform API. 
I understand your concern and agree that waiting for some reasonable time instead of forever is more suitable here.
@alisenchung Can you open a bug to replace forever waiting with timeout.
